File: src/managers.rs

pub struct PackageManagerInfo {
    pub name: &'static str,
    lockfiles: &'static [&'static str],
}

const PACKAGE_MANAGERS: &[PackageManagerInfo] = &[
    PackageManagerInfo {
        name: "npm",
        lockfiles: &["package-lock.json"],
    },
    PackageManagerInfo {
        name: "yarn",
        lockfiles: &["yarn.lock"],
    },
    PackageManagerInfo {
        name: "pnpm",
        lockfiles: &["pnpm-lock.yaml"],
    },
];

pub fn get_from_lockfile(file: &str) -> Option<&'static PackageManagerInfo> {
    PACKAGE_MANAGERS
        .iter()
        .find(|pm| pm.lockfiles.iter().any(|&f| f == file))
}

pub fn list_package_manager_names() -> Vec<&'static str> {
    PACKAGE_MANAGERS.iter().map(|pm| pm.name).collect()
}

pub fn get_package_manager_info_by_name(name: &str) -> Option<&'static PackageManagerInfo> {
    PACKAGE_MANAGERS.iter().find(|pm| pm.name == name)
}

File: npm/src/managers.ts
/**
 * Package manager information and detection logic
 */

export interface PackageManagerInfo {
  name: string;
  lockfiles: readonly string[];
}

const PACKAGE_MANAGERS: readonly PackageManagerInfo[] = [
  {
    name: 'npm',
    lockfiles: ['package-lock.json'],
  },
  {
    name: 'yarn',
    lockfiles: ['yarn.lock'],
  },
  {
    name: 'pnpm',
    lockfiles: ['pnpm-lock.yaml'],
  },
];

/**
 * Gets all possible lockfile names
 * @returns Array of lockfile names
 */
export function getPossibleLockfiles(): string[] {
  return PACKAGE_MANAGERS.flatMap((pm) => pm.lockfiles);
}

/**
 * Gets package manager info by lockfile name
 * @param file - The lockfile name
 * @returns Package manager info if found
 */
export function getFromLockfile(file: string): PackageManagerInfo | undefined {
  return PACKAGE_MANAGERS.find((pm) => pm.lockfiles.includes(file));
}

/**
 * Gets list of all package manager names
 * @returns Array of package manager names
 */
export function listPackageManagerNames(): string[] {
  return PACKAGE_MANAGERS.map((pm) => pm.name);
}

/**
 * Gets package manager info by name
 * @param name - The package manager name
 * @returns Package manager info if found
 */
export function getPackageManagerInfoByName(
  name: string
): PackageManagerInfo | undefined {
  return PACKAGE_MANAGERS.find((pm) => pm.name === name);
}

